.gitlab-ci.yml

```diff
@@ -0,0 +1,198 @@
+# typically because we build on a set of dependencies per machine.
+#
+# REBASELINE:
+# Defines the default choice for updating the saved baseline results. By default
+# the baseline can only be updated from the master branch. This variable offers
+# the option to manually ask for rebaselining from another branch if necessary.
+#
+# MFEM_ALLOC_NAME:
+# On LLNL's quartz, there is only one allocation shared among jobs in order to
+# save time and resources. This allocation has to be uniquely named so that we
+# are sure to retrieve it.
+
+variables:
+    BUILD_ROOT: ${CI_BUILDS_DIR}/mfem_${CI_COMMIT_REF_SLUG}_${CI_PIPELINE_ID}
+    BUILD_PATH: ${CI_BUILDS_DIR}/mfem_${CI_COMMIT_REF_SLUG}_${CI_PIPELINE_ID}/${PLAT}/${TOOLCHAIN}
+    REBASELINE: "NO"
+    MFEM_ALLOC_NAME: mfem_ci_${CI_PIPELINE_ID}
+
+stages:
+    - precleanup
+    - setup
+    - allocate
+    - libs
+    - build
+    - test
+    - deallocate
+    - baseline_check
+    - baseline_publish
+    - postcleanup
+
+.precleanup:
+    stage: precleanup
+    variables:
+        GIT_STRATEGY: none
+    script:
+        - rm -rf ${BUILD_ROOT}/${PLAT}
+        - rm -f ${BUILD_ROOT}/mfem
+
+# This section is LLNL-specific
+.setup:
+    stage: setup
+    variables:
+        GIT_STRATEGY: none
+    script:
+        - mkdir -p ${BUILD_ROOT} && cd ${BUILD_ROOT}
+        - if [ ! -d "tpls" ]; then git clone ssh://git@mybitbucket.llnl.gov:7999/mfem/tpls.git; fi
+        - if [ ! -d "tests" ]; then git clone ssh://git@mybitbucket.llnl.gov:7999/mfem/tests.git; fi
+        - cd tpls && git pull && cd ..
+        - cd tests && git pull && cd ..
+
+.with_gcc_6_1_0:
+    variables:
+        TOOLCHAIN: gcc_6_1_0
+    before_script:
+        - module load gcc/6.1.0
+
+.with_gcc_4_9_3:
+    variables:
+        TOOLCHAIN: gcc_4_9_3
+    before_script:
+        - module load gcc/4.9.3
+
+.build_hypre:
+    variables:
+        EXEC_PREFIX: ""
+    stage: libs
+    variables:
+        GIT_STRATEGY: none
+    script:
+        - cd ${BUILD_ROOT}/tpls
+        - mkdir -p ${BUILD_PATH}
+        - make clean
+        - |
+          if [[ "${MPIEXEC}" == "srun" ]]
+          then
+              export JOB_ID=$(squeue -h --name=${MFEM_ALLOC_NAME} --format=%A)
+              export EXEC_PREFIX="srun --jobid=${JOB_ID}"
+          fi
+        - $EXEC_PREFIX make -j 4 hypre PREFIX=${BUILD_PATH}/hypre/src/ BUILD_DIR=${BUILD_PATH} LOG_DIR=${BUILD_PATH}
+
+.build_metis:
+    variables:
+        EXEC_PREFIX: ""
+    stage: libs
+    variables:
+        GIT_STRATEGY: none
+    script:
+        - cd ${BUILD_ROOT}/tpls
+        - mkdir -p ${BUILD_PATH}
+        - |
+          if [[ "${MPIEXEC}" == "srun" ]]
+          then
+              export JOB_ID=$(squeue -h --name=${MFEM_ALLOC_NAME} --format=%A)
+              export EXEC_PREFIX="srun --jobid=${JOB_ID}"
+          fi
+        - $EXEC_PREFIX make -j 4 ${METIS} PREFIX=${BUILD_PATH}/ BUILD_DIR=${BUILD_PATH} LOG_DIR=${BUILD_PATH} CC=cc CXX=c++
```

Comment:
Shouldn't the `-j 4` and the `-j 8` below be machine-dependent?

Reply:
Now configurable per machine